File: typescript/src/rules/suppressions.ts
import type { Snapshot } from "../repo/repo.js";
import { typeScriptSourcePath } from "./source-paths.js";
import type { Definition, Finding } from "./types.js";

// A bare lint or type suppression in production TypeScript code is a
// finding: directives must carry a description on the line or sit under a
// preceding explanatory comment. The scan is line-based and reports the
// first offense per file.
export function suppressionFindings(
  definition: Definition,
  snapshot: Snapshot
): readonly Finding[] {
  const findings: Finding[] = [];
  for (const file of snapshot.files.values()) {
    if (!typeScriptSourcePath(file.path)) {
      continue;
    }
    const line = bareSuppressionLine(file.content);
    if (line !== undefined) {
      findings.push(suppressionFinding(definition, file.path, line));
    }
  }
  return findings;
}

function suppressionFinding(definition: Definition, filePath: string, line: number): Finding {
  return {
    rule_id: definition.id,
    severity: definition.severity,
    path: filePath,
    message: `${definition.message} (line ${String(line)})`
  };
}

function bareSuppressionLine(content: string): number | undefined {
  let previousLineIsComment = false;
  for (const [index, line] of content.split("\n").entries()) {
    if (bareSuppressionDirective(line, previousLineIsComment)) {
      return index + 1;
    }
    previousLineIsComment = explanatoryComment(line);
  }
  return undefined;
}

const directiveMarkers = [
  "eslint-disable",
  "oxlint-disable",
  "@ts-expect-error",
  "@ts-ignore",
  "biome-ignore"
] as const;

type DirectiveMarker = (typeof directiveMarkers)[number];

function bareSuppressionDirective(line: string, previousLineIsComment: boolean): boolean {
  const comment = commentText(line);
  const marker = directiveMarkers.find((candidate) => comment.includes(candidate));
  if (marker === undefined || previousLineIsComment) {
    return false;
  }
  const rest = comment.slice(comment.indexOf(marker) + marker.length);
  return !justifiedOnLine(marker, rest);
}

// Suppression directives only take effect inside comments, so directive
// names in code or string literals are not findings. The scan tracks quote
// state so a marker inside a string or template literal stays invisible.
function commentText(line: string): string {
  let quote: string | undefined;
  for (let index = 0; index < line.length; index += 1) {
    const character = line[index] ?? "";
    if (quote !== undefined) {
      [quote, index] = insideQuote(quote, character, index);
      continue;
    }
    if (quoteCharacter(character)) {
      quote = character;
      continue;
    }
    if (commentStart(line, index)) {
      return line.slice(index);
    }
  }
  return "";
}

function insideQuote(
  quote: string,
  character: string,
  index: number
): [string | undefined, number] {
  if (character === "\\") {
    return [quote, index + 1];
  }
  return [character === quote ? undefined : quote, index];
}

function quoteCharacter(character: string): boolean {
  return character === '"' || character === "'" || character === "`";
}

function commentStart(line: string, index: number): boolean {
  return line[index] === "/" && (line[index + 1] === "/" || line[index + 1] === "*");
}

function justifiedOnLine(marker: DirectiveMarker, rest: string): boolean {
  switch (marker) {
    case "@ts-ignore":
      return false;
    case "eslint-disable":
    case "oxlint-disable":
      return /\s--\s+\S/u.test(rest);
    case "biome-ignore":
      return /:\s*\S/u.test(descriptionText(rest));
    case "@ts-expect-error":
      return /\S/u.test(descriptionText(rest).replace(/^[\s:-]+/u, ""));
  }
}

function descriptionText(rest: string): string {
  return rest.replace(/\*\/.*$/u, "").trim();
}

// explanatoryComment treats a preceding comment line as justification, but
// never a comment that is itself a suppression directive.
function explanatoryComment(line: string): boolean {
  const trimmed = line.trimStart();
  const comment = trimmed.startsWith("//") || trimmed.startsWith("/*") || trimmed.startsWith("*");
  return comment && !directiveMarkers.some((marker) => trimmed.includes(marker));
}

File: python/src/slophammer_py/rules/suppressions.py
"""Suppression discipline for production Python code.

Suppression directives (noqa, type-ignore, and ty-ignore comments) need a
stated reason: trailing ``-- reason`` text or a preceding explanatory
comment. A type-ignore or ty-ignore without an error code is itself a
finding. Comments are located with the stdlib tokenizer, so directive text
inside strings and docstrings stays invisible; files the tokenizer rejects
fall back to a quote-tracking line scan. Test scope is exempt. The scan
reports the first offense per file.
"""

from __future__ import annotations

import io
import re
import tokenize

from ..core import Finding
from ..repo import RepoFile, Snapshot
from .definitions import Definition

NOQA = re.compile(r"#\s*noqa(?P<codes>:\s*[\w, ]+)?", re.IGNORECASE)
TYPE_IGNORE = re.compile(r"#\s*type:\s*ignore(?P<codes>\[[\w,\s-]+\])?")
TY_IGNORE = re.compile(r"#\s*ty:\s*ignore(?P<codes>\[[\w,\s-]+\])?")
REASON_TEXT = re.compile(r"--\s*\S")

EXEMPT_SEGMENTS = {"tests", "test", "fixtures", "templates", "testdata", "vendor", "scripts"}


def suppression_findings(definition: Definition, snapshot: Snapshot) -> list[Finding]:
    findings: list[Finding] = []
    for file in snapshot.files.values():
        if not production_python_path(file.path):
            continue
        line = bare_suppression_line(file)
        if line is not None:
            findings.append(
                Finding(
                    rule_id=definition.id,
                    severity=definition.severity,
                    path=file.path,
                    message=f"{definition.message} (line {line})",
                )
            )
    return findings


def production_python_path(path: str) -> bool:
    if not path.endswith(".py"):
        return False
    base = path.rsplit("/", 1)[-1]
    if base.startswith("test_") or base.endswith("_test.py") or base == "conftest.py":
        return False
    segments = set(path.split("/")[:-1])
    return not (segments & EXEMPT_SEGMENTS) and "migrations" not in segments


def bare_suppression_line(file: RepoFile) -> int | None:
    comments = comment_lines(file.content)
    explanatory = {line for line, text in comments.items() if explanatory_comment(text)}
    for line in sorted(comments):
        if bare_suppression_directive(comments[line], line - 1 in explanatory):
            return line
    return None


# Comment text by line number, via the tokenizer when the file parses and a
# quote-tracking line scan when it does not.
def comment_lines(content: str) -> dict[int, str]:
    try:
        return {
            item.start[0]: item.string
            for item in tokenize.generate_tokens(io.StringIO(content).readline)
            if item.type == tokenize.COMMENT
        }
    except (tokenize.TokenError, SyntaxError, ValueError):
        return fallback_comment_lines(content)


def fallback_comment_lines(content: str) -> dict[int, str]:
    comments: dict[int, str] = {}
    for index, line in enumerate(content.split("\n")):
        text = comment_text(line)
        if text:
            comments[index + 1] = text
    return comments


def bare_suppression_directive(comment: str, has_explanation: bool) -> bool:
    for pattern, codes_required in ((TYPE_IGNORE, True), (TY_IGNORE, True), (NOQA, False)):
        match = pattern.search(comment)
        if match is None:
            continue
        if codes_required and match.group("codes") is None:
            return True
        rest = comment[match.end() :]
        return not has_explanation and not REASON_TEXT.search(rest)
    return False


def explanatory_comment(text: str) -> bool:
    return not any(pattern.search(text) for pattern in (NOQA, TYPE_IGNORE, TY_IGNORE))


def comment_text(line: str) -> str:
    """The line's comment, starting at the first # outside string literals."""
    quote = ""
    index = 0
    while index < len(line):
        character = line[index]
        if quote:
            if character == "\\":
                index += 2
                continue
            if character == quote:
                quote = ""
        elif character in {'"', "'"}:
            quote = character
        elif character == "#":
            return line[index:]
        index += 1
    return ""
